src/main/groovy/lang/IntRange.java

```diff
@@ -377,40 +382,52 @@ public boolean containsAll(Collection other) {
 
     @Override
     public void step(int step, Closure closure) {
-        if (step == 0) {
-            if (!getFrom().equals(getTo())) {
+        step((Number) step, closure);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void step(Number step, Closure closure) {
+        final Long from = getFrom().longValue();
+        final Long to = getTo().longValue();
+
+        if (compareEqual(step, 0)) {
+            if (compareNotEqual(from, to)) {
                 throw new GroovyRuntimeException("Infinite loop detected due to step size of 0");
             } else {
                 return; // from == to and step == 0, nothing to do, so return
             }
         }
 
-        if (isReverse()) {
-            step = -step;
+        boolean isAscending = !isReverse();
+        if (compareLessThan(step, 0)) {
+            step = multiply(step, -1);
+            isAscending = !isAscending;
```

Comment:
github lists commit in an arbitrary order, but each commit is shown correctly by github, I think.